Client/WebFilters/ContentBlocker/ContentBlockerManager.swift

```diff
@@ -0,0 +1,457 @@
+    private(set) var compiledResources: [String: Resource]
+    
+    var pendingResources: [String: Resource] {
+      var results: [String: Resource] = [:]
+      
+      enabledResources.forEach { key, resource in
+        guard compiledResources[key] == nil || compiledResources[key] != resource else { return }
+        results[key] = resource
+      }
+      
+      return results
+    }
+    
+    init() {
+      self.enabledResources = [:]
+      self.compiledResources = [:]
+    }
+    
+    var isSynced: Bool {
+      return pendingResources.isEmpty
+    }
+    
+    /// Set a resource for the given rule type
+    func set(enabledResource: Resource, for ruleType: BlocklistRuleType) {
+      enabledResources[ruleType.identifier] = enabledResource
+    }
+    
+    /// Remove all resources for the given rule and source type
+    func removeEnabledResource(for ruleType: BlocklistRuleType) {
+      enabledResources.removeValue(forKey: ruleType.identifier)
+    }
+    
+    /// Remove all resources for the given rule and source type
+    func movePendingResource(forIdentifier identifier: String) {
+      compiledResources[identifier] = enabledResources[identifier]
+    }
+  }
+  
+  public static var shared = ContentBlockerManager()
+  /// The store in which these rule lists should be compiled
+  let ruleStore: WKContentRuleListStore
+  /// Compile results
+  private var cachedCompileResults: [String: (sourceType: BlocklistSourceType, result: Result<WKContentRuleList, Error>)]
+  /// The actor in which all of our sync data is stored on
+  private var data: SyncData
+  
+  /// A timer that runs which continually compiles new resources if it is needed
+  private var timer: Timer?
+  
+  init(ruleStore: WKContentRuleListStore = .default()) {
+    self.ruleStore = ruleStore
+    self.cachedCompileResults = [:]
+    self.data = SyncData()
+  }
+  
+  public func loadCachedCompileResults() async {
+    guard let identifiers = await ruleStore.availableIdentifiers() else { return }
+    let loadedCachedResults = await load(identifiers: identifiers)
+    
+    await Task { @MainActor in
+      for cachedResults in loadedCachedResults {
+        self.cachedCompileResults[cachedResults.0] = (.bundled, cachedResults.1)
+      }
+    }.value
+  }
+  
+  private func cleanupRuleLists() async {
+    guard let identifiers = await ruleStore.availableIdentifiers() else { return }
+    let enabledResources = await data.enabledResources
+    
+    await remove(identifiers: identifiers.filter({ identifier in
+      !enabledResources.contains(where: { $0.key == identifier })
+    }))
+  }
+  
+  private func remove(identifiers: [String]) async {
+    return await withTaskGroup(of: Void.self) { group in
+      for identifier in identifiers {
+        group.addTask {
+          do {
+            try await self.ruleStore.removeContentRuleList(forIdentifier: identifier)
+          } catch {
+            log.error(error)
```

Comment:
When unselecting any filter I get an error:
`2022-08-30 15:35:51.158 [Error] [ContentBlockerManager.swift:160] remove(identifiers:) > Error Domain=WKErrorDomain Code=8 "(null)" UserInfo={NSHelpAnchor=Rule list removal failed: Unspecified error during remove.}`

When selecting the same filter again Brave will send a HTTP request with a If-None-Match header, and the server returns 304 because the file didn't change since the last fetch. It looks like after selecting a filter again it doesn't properly load, because I can't see it loaded in this test:
```
TEST ======== AdblockEngineManager
TEST Loaded 7 (total) engine resources:
```

Tested on iOS 16

Reply:
Unselecting it a second time doesn't produce any errors but doesn't delete it from the device either.

Reply:
Thank you for the detailed error. I will look into this.

Reply:
@stoletheminerals This should be solved